tools/schemapi/schemapi.py

```diff
@@ -334,6 +384,109 @@ def _get_leaves_of_error_tree(
     return leaves
 
 
+def _lazy_group_tree_leaves(
+    errors: Iterable[ValidationError], /
+) -> Iterator[tuple[str, ValidationError]]:
+    """
+    Combines 3 previously distinct steps:
+
+    - ``_get_leaves_of_error_tree``
+    - (part of) ``_group_errors_by_json_path``
+        - Doesnt actually group yet, can by calling `dict(result)`.
+    - ``_is_required_value_error``
+    """  # noqa: D400
+    for err in errors:
+        if err_context := err.context:
+            yield from _lazy_group_tree_leaves(err_context)
+        elif err.validator == "required" and err.validator_value == ["value"]:
+            continue
+        else:
+            yield _json_path(err), err
+
+
+_fn_path = cast("Callable[[tuple[str, ValidationError]], str]", operator.itemgetter(0))
+"""Key function for ``(json_path, ValidationError)``."""
+_fn_validator = cast("Callable[[ValidationError], str]", operator.attrgetter("validator"))  # fmt: off
+"""Key function for ``ValidationError.validator``."""
+
+
+def _lazy_subset_to_most_specific_json_paths(
+    json_path_errors: Iterator[tuple[str, ValidationError]], /
+) -> Iterator[Iterable[ValidationError]]:
+    """
+    Currently using a `list`, but typing it more restrictive to see if it can be avoided.
+
+    - Needs to be sorted to work with groupby
+    - Reversing allows prioritising more specific groups, since they are seen first
+    - Then re-reversed, to keep seen order
+
+    """
+    rev_sort = sorted(json_path_errors, key=_fn_path, reverse=True)
+    keeping: dict[str, Iterable[ValidationError]] = {}
+    for unique_path, grouped_errors in groupby(rev_sort, key=_fn_path):
+        if any(seen.startswith(unique_path) for seen in keeping):
+            continue
+        else:
+            keeping[unique_path] = [err for _, err in grouped_errors]
+    yield from reversed(keeping.values())
+
+
+def _lazy_deduplicate_errors(
+    grouped_errors: Iterator[Iterable[ValidationError]], /
+) -> Iterator[ValidationError]:
+    for element_errors in grouped_errors:
+        for validator, errors in groupby(
+            sorted(element_errors, key=_fn_validator), key=_fn_validator
+        ):
+            if validator == "additionalProperties":
+                errors = _lazy_additional_properties(errors)
+            elif validator == "enum":
+                errors = _lazy_enum(errors)
+            yield from _lazy_unique_message(errors)
+
+
+def _lazy_unique_message(
+    iterable: Iterable[ValidationError], /
+) -> Iterator[ValidationError]:
+    seen = set()
+    for el in iterable:
+        if el.message not in seen:
+            seen.add(el.message)
+            yield el
+
+
+def _lazy_additional_properties(
+    iterable: Iterable[ValidationError], /
+) -> Iterator[ValidationError]:
+    it = iter(iterable)
+    first = next(it)
+    if (
+        parent := cast("ValidationError", first.parent)
+    ) and parent.validator == "anyOf":
+        yield min(_rechain(first, it), key=lambda x: len(x.message))
```

Comment:
TODO: rewrite this lambda

Reply:
Fixed in [`c261cb4` (#3547)](https://github.com/vega/altair/pull/3547/commits/c261cb46510ce680d99930bfc9a9f1f4762d1d9b)

See the last *"rules"* list in https://docs.python.org/3/howto/functional.html#small-functions-and-the-lambda-expression